mlir/lib/Quantum/Transforms/ChainedSelfInversePatterns.cpp

```diff
@@ -41,34 +43,176 @@ struct ChainedHadamardOpRewritePattern : public mlir::OpRewritePattern<CustomOp>
         LLVM_DEBUG(dbgs() << "Simplifying the following operation:\n" << op << "\n");
 
         StringRef OpGateName = op.getGateName();
-        if (!HermitianOps.contains(OpGateName))
+        if (!HermitianOps.contains(OpGateName)) {
             return failure();
+        }
 
         ValueRange InQubits = op.getInQubits();
         auto ParentOp = dyn_cast_or_null<CustomOp>(InQubits[0].getDefiningOp());
-        if (!ParentOp || ParentOp.getGateName() != OpGateName)
+        if (!ParentOp || ParentOp.getGateName() != OpGateName) {
             return failure();
+        }
 
         ValueRange ParentOutQubits = ParentOp.getOutQubits();
         // Check if the input qubits to the current operation match the output qubits of the parent.
         for (const auto &[Idx, Qubit] : llvm::enumerate(InQubits)) {
-            if (Qubit.getDefiningOp<CustomOp>() != ParentOp || Qubit != ParentOutQubits[Idx])
+            if (Qubit.getDefiningOp<CustomOp>() != ParentOp || Qubit != ParentOutQubits[Idx]) {
                 return failure();
+            }
         }
         ValueRange simplifiedVal = ParentOp.getInQubits();
         rewriter.replaceOp(op, simplifiedVal);
         return success();
     }
 };
 
+template <typename OpType>
+struct ChainedUUadjOpRewritePattern : public mlir::OpRewritePattern<OpType> {
+    using mlir::OpRewritePattern<OpType>::OpRewritePattern;
+
+    bool verifyParentGateType(OpType op, OpType parentOp) const
+    {
+        // Verify that the parent gate is of the same type.
+        // If OpType is quantum.custom, also verify that parent gate has the
+        // same gate name.
+
+        if (!parentOp || !isa<OpType>(parentOp)) {
+            return false;
+        }
+
+        if (isa<CustomOp>(op)) {
+            StringRef opGateName = cast<CustomOp>(op).getGateName();
+            StringRef parentGateName = cast<CustomOp>(parentOp).getGateName();
+            if (opGateName != parentGateName) {
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    bool verifyAllInQubits(OpType op, OpType parentOp) const
```

Comment:
Talked with @rmoyard , a lot of these verifications are common between cancel inverse and merge rotation, e.g. 
- checking the neighbouring ops are the same type (two quantum.custom, as opposed to one quantum.custom and one quantum.unitary)
- checking the first gate's output and the second gate's input are in the same order
- checking the ctrl qubits and values are shared

It might be worth it to pull them out into either tablegen or an analysis pass.

Reply:
https://github.com/PennyLaneAI/catalyst/pull/1186/commits/b9acfe69e7a5a9991f029b6df8d51de073aa142b